### PR TITLE
DRIVERS-3472 enforce client-side maximum SCRAM iteration count

### PR DESCRIPTION
**Pull request by connorsmacd**
Please complete the following before merging:

- [x] Is the relevant DRIVERS ticket in the PR title?
- [x] Update changelog.
- [x] Test changes in at least one language driver. [C Driver](https://github.com/mongodb/mongo-c-driver/pull/2332).
- [x] Test these changes against all server versions and topologies (including standalone, replica set, and sharded
    clusters).

# Summary

This PR makes changes to the Auth Spec to enforce a maximum client-side SCRAM iteration count. The changes here closely match what was suggested in the JIRA issue.

* Add URI option `maxScramIterations`.
  - **Default value of 100000**: This is ultimately arbitrary, but the 100K value suggested in the JIRA issue is the same value proposed for the server. I also found 100K to be a common choice for other projects that support SCRAM, e.g., [node-postgres](https://github.com/brianc/node-postgres/blob/1a38e1d774fd77a5f2ab37baa2d35720a1146672/packages/pg/lib/crypto/sasl.js#L33).
  - Ignored if less than 4096: Since the minimum iteration count is 4096, in only makes sense to mandate the maximum iteration count be greater than or equal to 4096.
* Update Auth Spec to enforce a maximum iteration count with testing requirements similar to the what was done for the minimum iteration count.

### CODE REVIEW

File: source/uri-options/uri-options.md [hunk 1]
```diff
@@ -78,6 +78,7 @@ to URI options apply here.
...
+| maxScramIterations                   | integer greater than or equal to 4096                                                                                                                                                     | 100000                                                                                                                                                                                                                       | no                                                                                                                                                                                                        | The maximum number of iterations permitted in SCRAM PBKDF2                                                                                                                                                                                                                                       |
```
Comment (kevinAlbs) [inline source/uri-options/uri-options.md:81]: Suggest adding "and less than or equal to 100000" in the "Accepted Values" description.

File: source/uri-options/tests/auth-options.yml [hunk 2]
```diff
@@ -23,3 +23,20 @@ tests:
...
+        warning: false
+        hosts: ~
+        auth: ~
+        options:
+            maxScramIterations: 123456
+    -
+        description: "maxScramIterations with value less than 4096 causes a warning"
+        uri: "mongodb://example.com/?maxScramIterations=4095"
+        valid: true
+        warning: true
```
Comment (kevinAlbs) [inline source/uri-options/tests/auth-options.yml:48]: Suggest testing boundary values to help avoid off-by-one issues in driver implementations:
- ~~100000 is OK, 100001 warns.~~ Edit: 100001 is not expected to warn.
- 4095 warns. 4096 is OK.

File: source/auth/auth.md [hunk 3]
```diff
@@ -634,8 +634,12 @@ that this be as secure and truly random as possible. For instance, Java provides
 SecureRandom class. SecureRandom is cryptographically generated while Random is just a pseudo-random generator with
 predictable outcomes.
 
-Additionally, drivers MUST enforce a minimum iteration count of 4096 and MUST error if the authentication conversation
-specifies a lower count. This mitigates downgrade attacks by a man-in-the-middle attacker.
+Drivers MUST enforce a minimum iteration count of 4096 and MUST error if the authentication conversation specifies a
+lower count. This mitigates downgrade attacks by a man-in-the-middle attacker.
+
+Drivers MUST enforce a maximum iteration count, defined by the `maxScramIterations` connection string option (default:
```
Comment (kevinAlbs) [inline source/auth/auth.md:640]: Opinion: I would rather avoid a new URI option unless there is an expected need. I expect the vast majority of users will not need to set `maxScramIterations`. The iteration count appears not configurable in Atlas.

OTOH the server changes also appear to make this value configurable: https://github.com/10gen/mongo/pull/55863. Consider asking in `#server-security` if there is any known need for iteration counts beyond, say, 100000 (or maybe some other high arbitrary limit). If there is no known need, I would be more inclined to drop the URI option.

File: source/uri-options/uri-options.md (hunk 1 above)
Comment (kevinAlbs) [reply source/uri-options/uri-options.md:81]: Doh, I misinterpreted this. This is expected to accept values beyond 100000. Resolving.

File: source/auth/auth.md (hunk 3 above)
Comment (connorsmacd) [reply source/auth/auth.md:640]: I agree the URI option is unlikely to be needed by most users. However, I felt it was still appropriate to provide the URI option for a few reasons:

- Existing systems that expect higher iteration counts would otherwise have no reasonable workaround.
- The defensive strength of an iteration count depends on the HMAC function. For example, in the context of password storage, [OWASP recommends 1.4M iterations for SHA1, but only 600K iterations for SHA256](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2).
- This is more of a philosophical issue, but the idea of drivers dictating what is a "high enough" iteration count feels perhaps overly opinionated.

If we do decide to get rid of the URI option in favor of a hard cap, then we may want to consider revising the number. I don't know how much the recommended values for *password storage* apply here, but the [OWASP cheatsheet](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2) I mentioned above is at least some hint that 100K may not be sufficiently high for a hard cap.

Comment (kevinAlbs) [reply source/auth/auth.md:640]: > #server-security

See [slack thread](https://mongodb.slack.com/archives/CB3CW8M8C/p1783346090178119).

> Existing systems that expect higher iteration counts would otherwise have no reasonable workaround.

Agreed. A too-low cap risks preventing auth with no workaround (very bad). A too-high cap might not address the original issue, and might not be practical to evaluate.

Plus, recommendations vary by time and standard. From CLAUDE:
> [10k](https://github.com/OWASP/CheatSheetSeries/blob/f7e6532fd/cheatsheets/Password_Storage_Cheat_Sheet.md#L73) (2019-01) => [500k](https://github.com/OWASP/CheatSheetSeries/blob/c68280f63/cheatsheets/Password_Storage_Cheat_Sheet.md#L66) (2019-05) => [310k](https://github.com/OWASP/CheatSheetSeries/blob/8253f2e37/cheatsheets/Password_Storage_Cheat_Sheet.md#L136) (2021-03) => [600k](https://github.com/OWASP/CheatSheetSeries/blob/ba94e5cbd/cheatsheets/Password_Storage_Cheat_Sheet.md#L142) (2023-01)

And from [NIST Special Publication 800-132](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-132.pdf):
> an iteration count of 10,000,000 may be appropriate

After further wavering: I am in favor of an option rather than a non-configurable cap. Regardless: I'd request adding to the "Q & A" section to document the rationale.

File: source/auth/auth.md [hunk 4]
```diff
@@ -2034,6 +2042,12 @@ See the speculative authentication section in the [MongoDB Handshake spec](../mo
 For SCRAM-SHA-1 and SCRAM-SHA-256, test that the minimum iteration count is respected. This may be done via unit testing
 of an underlying SCRAM library.
 
+### Maximum iteration count
```
Comment (kevinAlbs) [inline source/auth/auth.md:2045]: This test is similar to the "Minimum iteration count". I expect the "Minimum iteration count" cannot be tested as easily end-to-end, since the server rejects attempts to set too-low iteration counts:

```python
client["admin"].command({"setParameter": 1, "scramSHA256IterationCount": 1000})
# Server error: "1000 is not greater than or equal to 5000"
```

But I expect testing higher iteration counts are rejected by the client should be possible end-to-end. Suggest replacing this test with an end-to-end test like [the following](https://gist.github.com/kevinAlbs/2500920c69a4c7cb8be84b9301d14a0e):

```python
# Test SCRAM-SHA-256:
scramSHA256IterationCount = test_client()["admin"].command({"getParameter": 1, "scramSHA256IterationCount": 1})["scramSHA256IterationCount"]
expect_error(test_client(mech="SCRAM-SHA-256", max_iterations=scramSHA256IterationCount - 1))
expect_ok(test_client(mech="SCRAM-SHA-256", max_iterations=scramSHA256IterationCount))

# Test SCRAM-SHA-1:
scramIterationCount = test_client()["admin"].command({"getParameter": 1, "scramIterationCount": 1})["scramIterationCount"]
expect_error(test_client(mech="SCRAM-SHA-1", max_iterations=scramIterationCount - 1))
expect_ok(test_client(mech="SCRAM-SHA-1", max_iterations=scramIterationCount))
```

File: source/auth/auth.md [hunk 5]
```diff
@@ -634,8 +634,12 @@ that this be as secure and truly random as possible. For instance, Java provides
 SecureRandom class. SecureRandom is cryptographically generated while Random is just a pseudo-random generator with
 predictable outcomes.
 
-Additionally, drivers MUST enforce a minimum iteration count of 4096 and MUST error if the authentication conversation
-specifies a lower count. This mitigates downgrade attacks by a man-in-the-middle attacker.
+Drivers MUST enforce a minimum iteration count of 4096 and MUST error if the authentication conversation specifies a
+lower count. This mitigates downgrade attacks by a man-in-the-middle attacker.
+
+Drivers MUST enforce a maximum iteration count, defined by the `maxScramIterations` connection string option (default:
+100000), and MUST error if the authentication conversation specifies a higher count. This mitigates client-side denial
```
Comment (kevinAlbs) [inline source/auth/auth.md:641]:
```suggestion
600000), and MUST error if the authentication conversation specifies a higher count. This mitigates client-side denial
```

Suggest using a default to match the current [OWASP recommendation](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html):

> If FIPS-140 compliance is required, use PBKDF2 with a work factor of 600,000 or more and set with an internal hash function of HMAC-SHA-256.

MongoDB also documents [FIPS-140 compliance](https://www.mongodb.com/docs/manual/tutorial/configure-fips/) instructions. So this may reduce the chance of users receiving errors on upgrade (and 600,000 still seems low enough to not be a cause of concern).

File: source/auth/auth.md (hunk 3 above)
Comment (jyemin) [inline source/auth/auth.md:640]: There is an existing technique for adding mechanism-specific configuration: mechanism_properties.  We use it for other mechanisms, but not yet for SCRAM.  It would look something like this:

```
mongodb://jeff:pwd@localhost?authMechanismProperties=MAX_ITERATIONS:200000
```

Can we use this instead of adding a top-level parameter?

**Review by jyemin — CHANGES_REQUESTED** [review]
